File: 2020/1570.py
"""
1570
"""
import collections
class SparseVector:
    # Get TLE if use bitwise operation...
    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.nd = {}
        for i in range(len(nums)-1,-1,-1):
            if nums[i] != 0:
                self.nd[i] = nums[i]

    # Return the dotProduct of two sparse vectors
    def dotProduct(self, vec):
        """
        :type vec: 'SparseVector'
        :rtype: int
        """
        res = 0
        for k, v in self.nd.items():
            if k in vec.nd.keys():
                res += v*vec.nd[k]
        return res

    
# Your SparseVector object will be instantiated and called as such:
# v1 = SparseVector(nums1)
# v2 = SparseVector(nums2)
# ans = v1.dotProduct(v2)

File: current_session/c++/1570.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class SparseVector {
public:
  map<int,int> mp;
  int size;
  SparseVector(vector<int> &nums) {
    for (int i=0; i<nums.size();i++) {
      if (nums[i]!=0) mp[i] = nums[i];
    }
    size = nums.size();
  }

  // Return the dotProduct of two sparse vectors
  int dotProduct(SparseVector& vec) {
    int res=0;
    for (int i=0; i<size; i++) {
      if (mp.find(i) != mp.end() && vec.mp.find(i) != vec.mp.end()) res += mp[i]*vec.mp[i];
    }
    return res;
  }
};

// Your SparseVector object will be instantiated and called as such:
// SparseVector v1(nums1);
// SparseVector v2(nums2);
// int ans = v1.dotProduct(v2);